### PR TITLE
feat: introduce migrate command

### PR DESCRIPTION
**Pull request by KristianDD**
## PR Checklist

- [ ] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [ ] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [ ] You have signed the [CLA](http://www.nativescript.org/cla).
- [ ] All existing tests are passing: https://github.com/NativeScript/nativescript-cli/blob/master/CONTRIBUTING.md#contribute-to-the-code-base
- [ ] Tests for the changes are included.

## What is the new behavior?

Fixes/Implements/Closes #[Issue Number].

### CODE REVIEW

File: lib/controllers/migrate-controller.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+			const coerceMaxSatisfying = semver.coerce(maxSatisfyingVersion).version;
+			const coerceVerifiedVersion = semver.coerce(dependency.verifiedVersion).version;
+			//This makes sure that if the user has a prerelease 6.0.0-next-2019-06-10-092158-01 version we will update it to 6.0.0
+			if (isPrereleaseVersion) {
+				if (semver.gt(coerceMaxSatisfying, coerceVerifiedVersion)) {
+					return false;
+				}
+
+				//TODO This should be removed once we update the verified versions to no prerelease versions
+				if (isPrereleaseVersion && semver.eq(maxSatisfyingVersion, dependency.verifiedVersion)) {
```
Comment (KristianDD) [inline lib/controllers/migrate-controller.ts]: remove isPrereleaseVersion

File: lib/package-installation-manager.ts [hunk 2]
```diff
@@ -33,10 +33,14 @@ export class PackageInstallationManager implements IPackageInstallationManager {
...
 
+		return await this.maxSatisfyingVersion(packageName, compatibleVersionRange) || latestVersion;
+	}
+
+	public async maxSatisfyingVersion(packageName: string, versionRange: string): Promise<string> {
 		const data = await this.$packageManager.view(packageName, { "versions": true });
 
-		const maxSatisfying = semver.maxSatisfying(data, compatibleVersionRange);
-		return maxSatisfying || latestVersion;
+		const maxSatisfying = semver.maxSatisfying(data, versionRange);
```
Comment (KristianDD) [inline lib/package-installation-manager.ts]: return the result right away

File: lib/controllers/migrate-controller.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+		this.$logger.info("Migration complete.");
+	}
+
+	private async shouldMigrateDependencyVersion(dependency: IMigrationDependency, projectData: IProjectData): Promise<boolean> {
+		const collection = dependency.isDev ? projectData.devDependencies : projectData.dependencies;
+		if (
+			collection &&
+			collection[dependency.packageName] &&
+			(semver.valid(collection[dependency.packageName]) || semver.validRange(collection[dependency.packageName]))
```
Comment (KristianDD) [inline lib/controllers/migrate-controller.ts]: Maybe should handle tags if not a valid version or range.

File: lib/controllers/migrate-controller.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+		this.$fs.deleteFile(path.join(projectData.projectDir, constants.PACKAGE_LOCK_JSON_FILE_NAME));
+		this.$fs.deleteFile(path.join(projectData.projectDir, constants.TSCCONFIG_TNS_JSON_NAME));
+		this.$logger.info("Clean old project artefacts complete.");
+	}
+
+	private async migrateDependencies(projectData: IProjectData): Promise<void> {
+		this.$logger.info("Start migrating dependencies.");
+		for (let i = 0; i < MigrateController.migrationDependencies.length; i++) {
+			const dependency = MigrateController.migrationDependencies[i];
+			if (this.shouldSkipMigrationDependency(dependency, projectData)) {
```
Comment (DimitarTachev) [inline lib/controllers/migrate-controller.ts]: consider unifying/reusing logic from shouldMigrate

File: lib/controllers/migrate-controller.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+				continue;
+			}
+
+			if (dependency.replaceWith) {
+				this.$pluginsService.removeFromPackageJson(dependency.packageName, dependency.isDev, projectData.projectDir);
+				const replacementDep = _.find(MigrateController.migrationDependencies, migrationPackage => migrationPackage.packageName === dependency.replaceWith);
```
Comment (DimitarTachev) [inline lib/controllers/migrate-controller.ts]: throw Error if undefined

File: lib/definitions/migrate.d.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+interface IDependency {
+	packageName: string;
+	isDev?: boolean;
+}
+
+interface IMigrationDependency extends IDependency {
+	mustRemove?: boolean;
+	replaceWith?: string;
+	verifiedVersion?: string;
```
Comment (DimitarTachev) [inline lib/definitions/migrate.d.ts]: verifiedVersion -> targetVersion
shouldAdd -> shouldAddIfMissing

File: lib/definitions/migrate.d.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+	shouldMigrate(data: IProjectDir): Promise<boolean>;
+}
+
+interface IDependency {
+	packageName: string;
+	isDev?: boolean;
+}
+
+interface IMigrationDependency extends IDependency {
+	mustRemove?: boolean;
```
Comment (DimitarTachev) [inline lib/definitions/migrate.d.ts]: mustRemove -> shouldRemove

File: lib/definitions/migrate.d.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
+interface IMigrateController {
+	migrate(migrateData: IProjectDir): Promise<void>;
+	shouldMigrate(data: IProjectDir): Promise<boolean>;
+}
+
+interface IDependency {
```
Comment (DimitarTachev) [inline lib/definitions/migrate.d.ts]: Consider using it in as a base interface

**Pull request closed without merging**